### DISCUSSION TITLE
Future Obsidian Features -- Discussion and WIP

### DISCUSSION THREAD

**Discussion by JPT62089** [tim-hub/obsidian-bible-reference#77] [Show and tell]
Hey @tim-hub and community...

I've been taking some time to actually learn how to write plugins for Obsidian and have basically been rewriting some things starting with the settings page... This re-write has solely Bolls Life API in mind and I'm hoping to do a bunch of things with it.

## 1. Rewrite settings page to use Svelte rather than the built-in Obsidian page.
This helps accomplish one main goal at this point. Support for multiple Language/Translations. Here's an example of what I've done so far. It's mostly 'complete' at this point, but still a lot of work to fine tune / refactoring before it's ready for prime.

https://user-images.githubusercontent.com/1808107/215902406-850dbb3c-2582-48e1-a2e0-4066fa62fc4e.mp4

The initial goal of this was to dynamically pull what languages and translations are supported from Bolls Life. But for now it's just a local JSON file from Bolls Life.

## 2. Support Localization
The next thing I've just started working on is implementing localization. One of the main packages that is used seems to be i18next so I've been using that. Here's a 5 minute implementation example of that at work.

https://user-images.githubusercontent.com/1808107/215903691-d9732b6b-5e50-4013-9cac-9fad1844fd78.mp4

## 3. Refactor / Rewrite the verse detection
This one I've barely started / looked at. The most I've done was have a little fun with [RegEx](https://regexr.com/76kiu) (Note: This does not work with books like 1/2 Corinthians yet). With this regex it should be able to detect and categorize each of the parts of the following. Tho like I said it's barely started and the above link is mostly all I've done :)

```
--john1:1
--john1:1-5
--john1:50-2:3
--esv:john1:1
--esv:john1:1-5
--esv:john1:50-2:3
--esv,nasb:john1:1
--esv,nasb:john1:1-5
--esv,nasb:john1:50-2:3
--john 1:1
--john 1:1-5
--john 1:50-2:3
--esv:john 1:1
--esv:john 1:1-5
--esv:john 1:50-2:3
--esv,nasb:john 1:1
--esv,nasb:john 1:1-5
--esv,nasb:john 1:50-2:3
-- john1:1
-- john1:1-5
-- john1:50-2:3
-- esv:john1:1
-- esv:john1:1-5
-- esv:john1:50-2:3
-- esv,nasb:john1:1
-- esv,nasb:john1:1-5
-- esv,nasb:john1:50-2:3
-- john 1:1
-- john 1:1-5
-- john 1:50-2:3
-- esv:john 1:1
-- esv:john 1:1-5
-- esv:john 1:50-2:3
-- esv,nasb:john 1:1
-- esv,nasb:john 1:1-5
-- esv,nasb:john 1:50-2:3
```

What are your thoughts / concerns / conundrums / snarks?

**Comment by tim-hub** [comment]
thanks @JPT62089 

I like the ideas.


of localization. and  Regex to match the bible verse

**Reply by tim-hub** [reply]
The only concern from me is to use Svelte instead of current build in page.

Because I am not very familiar with Svelte and the advantages and dis-advantages

**Reply by tim-hub** [reply]
I was reading this, 

It mentioned this 


> Svelte is built around a compiler that preprocesses your code and outputs vanilla JavaScript, which means it doesn't need to load any libraries at run time. This also means that it doesn't need a virtual DOM to track state changes, which allows your plugin to run with minimal additional overhead.


Which sounds cool and powerful for Obsidian plugin, I think I like the Svelte idea as well to remove some limitation of current plugin dev.

**Reply by JPT62089** [reply]
Yeah, I was trying to figure out a way to automatically update the fields so that we can better organize the translations based on language selected. I don't want this to be based on the language that the users system or Obsidian is set to. Default, sure, but not limited to... And with more and more translations being added it's becoming quite the list.

As I'm sure you noticed when you select the language it only shows that languages translations. You don't have to update the whole page in order to update one field. That's really kinda what sold me on it. Probably possible with some other frameworks, but the only one I found with [tutorials that directly addressed this with Obsidian plugins](https://www.youtube.com/watch?v=mCF80HBfUWA).

We can also use Svelte just as easily in a view on the sides that could be used for things like searching text within the Bible or getting verse references without using a command. Or just displaying the verses on the side rather than in your notes. Lots of things we can do but yeah... If you want to check out what I've done, here's a link to my [fork/branch](https://github.com/JPT62089/obsidian-bible-reference/tree/rewrite) for this. Bible references AFAIRemember are broken at this point but settings UI is working. Also lots of refactoring needing to be done :)

Edit: Included YT link for [Svelte+Obsdian tutorial](https://www.youtube.com/watch?v=mCF80HBfUWA)

**Comment by tim-hub** [comment]
From what I know so far, Svelte probably can help for the `caching` feature as well to cache the api response locally

**Reply by JPT62089** [reply]
Nice! We can definitely look into that in the future. Also, I've fallen off of this project due to work being a bit crazy but hopefully will be able to get back to it in the coming weeks.
